hydroshare_jupyter_sync/fs_event_handler.py

```diff
@@ -0,0 +1,92 @@
+from watchdog.events import (
+    FileSystemEventHandler,
+    PatternMatchingEventHandler,
+    FileCreatedEvent,
+    FileModifiedEvent,
+    FileDeletedEvent,
+    FileMovedEvent,
+    FileClosedEvent,
+)
+
+from .fs_events import Events
+from .lib.filesystem.fs_resource_map import LocalFSResourceMap
+from .lib.events.event_broker import EventBroker
+
+from functools import wraps
+import logging
+
+# module level log
+logger = logging.getLogger(__name__)
+
+
+def log_event(fn):
+    @wraps(fn)
+    def wrapper(self, event):
+        logger.info(event)
+        return fn(self, event)
+
+    return wrapper
+
+
+def fs_event_handler_factory(event_broker: EventBroker) -> FileSystemEventHandler:
+    """Wrap FSEventHandler in event_broker context."""
+
+    class FSEventHandler(PatternMatchingEventHandler):
+        def __init__(self, local_fs_map: LocalFSResourceMap):
+            # TODO: use pattern kwarg to ignore certain files/file extensions. it would be nice if
+            # this were a configurable.
+            super().__init__(ignore_directories=True)
+
+            # dependency inject local filesystem map
+            self._res_map = local_fs_map
+
+        @log_event
+        def on_any_event(self, event):
+            # log all events
+            ...
+
+        def on_created(self, event: FileCreatedEvent) -> None:
+            # add file to local fs map
+            self._res_map.add_file(event.src_path)
+
+            # dispatch new state
+            event_broker.dispatch(Events.STATUS, self.resource_id)
```

Comment:
The event broker dispatch could be a decorator like the log_event.

Reply:
I agree with you. Given that the dispatch is more or less just adding functionality that is more appropriate for decorator. I am going to leave this open if I have any implementation discussion to add.